src/main/scala/mesosphere/marathon/core/group/impl/GroupManagerImpl.scala

```diff
@@ -45,7 +45,16 @@ class GroupManagerImpl(
     */
   private[this] val root = LockedVar(initialRoot)
 
-  override def rootGroup(): RootGroup = root.get()
+  override def rootGroup(): RootGroup =
+    root.get() match { // linter:ignore:UseGetOrElseNotPatMatch
+      case None =>
+        val group = Await.result(groupRepository.root(), config.zkTimeoutDuration)
```

Comment:
There is no race condition because of some subtleties. It is called first by a leader when performing migration, which is sequential, and there is a synchronization point which is `serializeUpdates: WorkQueue`. But if considering this code not in the overall context, I do agree that a race condition is possible (assuming there is no sync point), though `val group = root := Some(Await.result(groupRepository.root(), config.zkTimeoutDuration))` has the same problem. I will update the code to handle this properly no matter how and when this method is invoked. Thanks a lot for pointing this out.

Reply:
You're right - one line suffers from the same problem. And `root.update()` seems to solve the problem. Thx.